mono/metadata/threadpool-ms-io.c

```diff
@@ -62,23 +68,42 @@ enum {
 	AIO_OP_LAST
 };
 
+typedef enum {
+	UPDATE_EMPTY = 0,
+	UPDATE_ADD,
+	UPDATE_REMOVE_SOCKET,
+	UPDATE_REMOVE_DOMAIN,
+} ThreadPoolIOUpdateType;
+
 typedef struct {
 	gint fd;
 	MonoSocketAsyncResult *sockares;
-} ThreadPoolIOUpdate;
+} ThreadPoolIOUpdate_Add;
```

Comment:
To fix that, a simple 'mono_cond_wait(&threadpool_io->updates_cond,
&threadpool_io->updates_lock)' at threadpool-ms-io.c:658 would do it, as
that would force the sockares to be added to the states table (which is
marked as a root) while it would still be present on the stack of the
mono_threadpool_ms_io_add function.

On Friday, August 21, 2015, Rodrigo Kumpera <notifications@github.com>
wrote:

> In mono/metadata/threadpool-ms-io.c
> <https://github.com/mono/mono/pull/1969#discussion_r37608758>:
>
> >  typedef struct {
> >  	gint fd;
> >  	MonoSocketAsyncResult *sockares;
> > -} ThreadPoolIOUpdate;
> > +} ThreadPoolIOUpdate_Add;
>
> This is broken, we're storing heap pointers without any safeguards. This
> will cause issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mono/mono/pull/1969/files#r37608758>.
>


-- 
Ludovic Henry
Runtime Software Engineer
Xamarin, Inc.